### PR TITLE
[FLINK-24697][flink-connectors-kafka] add auto.offset.reset configuration for group-offsets startup mode

### PR DESCRIPTION
**Pull request by ruanhang1993**
## What is the purpose of the change
 
This PR provides the way to change the 'auto.offset.reset' for kafka table source when use 'group-offsets' startup mode .

When 'auto.offset.reset' is set, the 'group-offsets' startup mode will use the provided auto offset reset strategy, or else 'none' reset strategy in order to be consistent with the DataStream API.

## Brief change log

  - When 'auto.offset.reset' is set, the 'group-offsets' startup mode will use the provided auto offset reset strategy, or else 'none' reset strategy as default

## Verifying this change

  - Added test that validates that the 'auto.offset.reset' is set for kafka consumers

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by hanwei-36** [comment]
When to merge this PR？

**Comment by ruanhang1993** [comment]
@fapaul Could you help to review this PR?
By the way, please assign this ticket to me in the JIRA.

**Review by ashulin — COMMENTED** [review]
I left some comments.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 1]
```diff
@@ -393,7 +396,16 @@ public int hashCode() {
                 kafkaSourceBuilder.setStartingOffsets(OffsetsInitializer.latest());
                 break;
             case GROUP_OFFSETS:
-                kafkaSourceBuilder.setStartingOffsets(OffsetsInitializer.committedOffsets());
+                String offsetResetConfig =
+                        properties.getProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG);
+                if (offsetResetConfig == null) {
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:401]: > Maybe you can get a default value `none` and avoid judging null.

The properties for the Kafka client are not from fixed table options. So we have no way to set a default value.

[the remark this answers is not in the document]
Comment (ashulin) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:401]: properties.getProperty(key, defaultValue);

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 2]
```diff
@@ -380,6 +381,44 @@ public void testTableSourceCommitOnCheckpointDisabled() {
...
+        testSetOffsetResetForStartFromGroupOffsets("latest");
+    }
+
+    private void testSetOffsetResetForStartFromGroupOffsets(String value) {
+        final Map<String, String> modifiedOptions =
+                getModifiedOptions(
+                        getBasicSourceOptions(),
+                        options -> {
+                            options.remove("scan.startup.mode");
+                            if (value != null) {
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:398]: The properties for the Kafka client are not from fixed table options. So we have no way to set a default value. And we have to test the null value.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:401]: Yeah, I get your point.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 3]
```diff
@@ -393,7 +396,16 @@ public int hashCode() {
...
+                    kafkaSourceBuilder.setStartingOffsets(OffsetsInitializer.committedOffsets());
+                } else {
+                    kafkaSourceBuilder.setStartingOffsets(
+                            OffsetsInitializer.committedOffsets(
+                                    OffsetResetStrategy.valueOf(
+                                            offsetResetConfig.toUpperCase(Locale.ROOT))));
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java]: fixed

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java]: fixed

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 4]
```diff
@@ -393,7 +396,27 @@ public int hashCode() {
...
+                                .filter(
+                                        ors ->
+                                                ors.name()
+                                                        .equals(
+                                                                offsetResetConfig.toUpperCase(
+                                                                        Locale.ROOT)))
+                                .findAny()
+                                .orElse(null);
+                Preconditions.checkArgument(
+                        offsetResetStrategy != null,
```
Comment (ashulin) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:414]: Can use `Preconditions.checkNotNull()`

Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:414]: > Can use `Preconditions.checkNotNull()`

I think this place should throw an IllegalArgumentException instead of a NPE, so I use the `Preconditions. checkArgument`. WDYT?

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by ruanhang1993** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 5]
```diff
@@ -393,7 +396,27 @@ public int hashCode() {
...
+                        Arrays.stream(OffsetResetStrategy.values())
+                                .filter(
+                                        ors ->
+                                                ors.name()
+                                                        .equals(
+                                                                offsetResetConfig.toUpperCase(
+                                                                        Locale.ROOT)))
+                                .findAny()
+                                .orElse(null);
+                Preconditions.checkArgument(
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:413]: Nit: you can immediately throw the `IllegalArgumentException` with `.orElseThrow()`

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 6]
```diff
@@ -393,7 +396,27 @@ public int hashCode() {
...
+                                ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
+                                OffsetResetStrategy.NONE.name());
+                OffsetResetStrategy offsetResetStrategy =
+                        Arrays.stream(OffsetResetStrategy.values())
+                                .filter(
+                                        ors ->
+                                                ors.name()
+                                                        .equals(
+                                                                offsetResetConfig.toUpperCase(
+                                                                        Locale.ROOT)))
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:410]: Nit: This is a bit hard to read maybe extract a method.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 7]
```diff
@@ -380,6 +389,53 @@ public void testTableSourceCommitOnCheckpointDisabled() {
                                 .noDefaultValue()));
     }
 
+    @ParameterizedTest
+    @MethodSource("offsetResetStrategy")
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:393]: I do not think the method source is necessary here you can use the following:

```java
    @ValueSource(strings = {"none", "earliest", "latest"})
    @NullSource
```

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 8]
```diff
@@ -380,6 +389,53 @@ public void testTableSourceCommitOnCheckpointDisabled() {
                                 .noDefaultValue()));
     }
 
+    @ParameterizedTest
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:392]: Can you also quickly migrate all the other tests to use junit 5 because mixing them in the same class causes problems? It would be nice to do it as a separate commit.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:415]: Nit: Personally, I am a big fan of handling all cases explicitly

```java
if (value == null) {
    return;
}
...
```
it makes the code often more readable

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 9]
```diff
@@ -380,6 +389,53 @@ public void testTableSourceCommitOnCheckpointDisabled() {
...
+                        options -> {
+                            options.remove("scan.startup.mode");
+                            if (value != null) {
+                                options.put(
+                                        PROPERTIES_PREFIX + ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
+                                        value);
+                            }
+                        });
+        final DynamicTableSource tableSource = createTableSource(SCHEMA, modifiedOptions);
+        assertThat(tableSource, instanceOf(KafkaDynamicSource.class));
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:422]: Please use for all the new assertions assertJ.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java [hunk 10]
```diff
@@ -821,6 +826,175 @@ public void testPerPartitionWatermarkWithIdleSource() throws Exception {
...
+                        "+I[0, 0]", "+I[0, 1]", "+I[0, 2]", "+I[1, 3]", "+I[1, 4]", "+I[1, 5]");
+        testStartFromGroupOffsets("earliest", expected);
+    }
+
+    @Test
+    public void testStartFromGroupOffsetsNone() {
+        try {
+            testStartFromGroupOffsetsWithException("none", Collections.emptyList());
+            fail("None offset reset error.");
+        } catch (Exception e) {
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:847]: Why do you not only catch the `ExecutionException`? Currently afaict the `fail` call is also caught and handeled.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java [hunk 11]
```diff
@@ -821,6 +826,175 @@ public void testPerPartitionWatermarkWithIdleSource() throws Exception {
...
+                        + ") WITH (\n"
+                        + "  'connector' = 'values'\n"
+                        + ")";
+        tEnv.executeSql(createSink);
+
+        TableResult tableResult =
+                tEnv.executeSql("INSERT INTO " + sinkName + " SELECT * FROM " + tableName);
+
+        if (expected.size() == 0) {
+            KafkaTableTestUtils.waitingEmptyResults(sinkName, Duration.ofSeconds(5));
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:917]: I am not sure if it is a good idea to introduce another test util for waiting. We already have `CommonTestUtils#waitUntil` can you take a look at that?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java [hunk 12]
```diff
@@ -821,6 +826,175 @@ public void testPerPartitionWatermarkWithIdleSource() throws Exception {
...
+        try {
+            testStartFromGroupOffsetsWithException("none", Collections.emptyList());
+            fail("None offset reset error.");
+        } catch (Exception e) {
+            Throwable rootCause = ExceptionUtils.getRootCause(e);
+            assertTrue(rootCause instanceof NoOffsetForPartitionException);
+        }
+    }
+
+    private void testStartFromGroupOffsets(String reset, List<String> expected)
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:853]: The two methods `testStartFromGroupOffset` and `testStartFromGroupOffset` look like largely duplicated code. Can you extract the common parts into a method?

**Review by fapaul — COMMENTED** [review]
@ruanhang1993 thanks for the fix this is definitely an important feature. The production code looks mostly good I left some nitty suggestions.

I left some more comments for the test code where some changes are needed.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java (hunk 5 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java]: fixed

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java (hunk 6 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java]: fixed

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java (hunk 7 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java]: fixed

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java (hunk 8 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:383]: fixed

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java (hunk 11 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java]: fixed. I have refactored them with the `CommonTestUtils#waitUntil`.
This code aims to make sure that we will not get any result from the test. I have reimplemented it with `CommonTestUtils#waitUntil` and give it a more meaningful name.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java (hunk 12 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java]: I have fixed it.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 13]
```diff
@@ -679,106 +729,165 @@ private void verifyEncoderSubject(
...
+                .satisfies(
+                        throwable ->
+                                assertThatChainOfCauses(throwable)
+                                        .anySatisfy(
+                                                cause ->
+                                                        assertThat(cause)
+                                                                .isInstanceOf(
+                                                                        ValidationException.class)
+                                                                .hasMessage(
+                                                                        "Option 'topic' and 'topic-pattern' shouldn't be set together.")));
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:754]: For the nested exceptions, you can use Flink's util `ExceptionUtils#findThrowable` or `FlinkAssertions.anyCauseMatches()`

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestUtils.java [hunk 14]
```diff
@@ -74,23 +77,40 @@
...
+    }
 
-        // timeout, assert again
-        List<String> actual = TestValuesTableFactory.getResults(sinkName);
-        Collections.sort(actual);
-        assertEquals(expected, actual);
+    public static void checkReturnEmptyResults(String sinkName, Duration timeout)
+            throws InterruptedException, TimeoutException {
+        CommonTestUtils.waitUtil(
+                new Supplier<Boolean>() {
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestUtils.java:96]: I am not the biggest fan of anonymous classes including complex logic, can you extract it? WDYT about adding a common helper to `CommonTestUtils` that combines `waitUntil` and your count-based logic?

**Review by fapaul — COMMENTED** [review]
Thanks for the update. It looks almost ready only two small points are left.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java (hunk 13 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:754]: `FlinkAssertions.anyCauseMatches()` uses `hasMessageContaining` to check the error message. We need to check with `hasMessage` here.
I will add a new method `containsCause` to `FlinkAssertions` to fix it.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestUtils.java (hunk 14 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestUtils.java]: Yes, I do not think waiting and checking the empty result is a good way to test this situation. I have changed the test to write some data after the job started, and check the new data to make sure we read from latest offsets.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 15]
```diff
@@ -417,6 +426,19 @@ public int hashCode() {
...
 
+    private OffsetResetStrategy getResetStrategy(String offsetResetConfig) {
+        return Arrays.stream(OffsetResetStrategy.values())
+                .filter(ors -> ors.name().equals(offsetResetConfig.toUpperCase(Locale.ROOT)))
+                .findAny()
+                .orElseThrow(
+                        () ->
+                                new IllegalArgumentException(
+                                        String.format(
+                                                "%s can not be set to %s. Valid values: [latest, earliest, none]",
```
[the remark this answers is not in the document]
Comment (OneCricketeer) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:437]: e.g. `Arrays.stream(OffsetResetStrategy.values()).map(Enum::name).map(String::toLowerCase())`

**Review by fapaul — APPROVED** [review]
LGTM now. Please rebase the changes to the latest master and squash the commits :)

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java]: fixed
